### PR TITLE
TensorBoard/Wandb/optuna/raytune integration improvements.

### PR DESCRIPTION
**Pull request by madlag**
Improves TensorBoard logging by grouping train / eval metrics as it is usually done in TensorBoard.
Improves TensorBoard/optuna model hyper-parameters logging.
Improves optuna and Ray/tune integration, and provides model hyper-parameter naming.

Test (and sample code) is provided in test_trainer.TrainerHyperParameterIntegrationTest .

Some more work may be need to harmonize metrics naming for eval / train, as the "eval_" prefix used is not very convenient, using a "eval/" prefix would be more foolproof, and consistent with TensorBoard usage, but it would break quite some code, and so may be done in a separate PR.

### CODE REVIEW

File: src/transformers/testing_utils.py [hunk 1]
```diff
@@ -233,6 +234,32 @@ def require_faiss(test_case):
         return test_case
 
 
+def require_optuna(test_case):
```
Comment (sgugger) [inline src/transformers/testing_utils.py:237]: Note that we don't have any setup installing optuna or ray on the CI so for now, any tests marked by those decorators is just never going to be executed.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -702,6 +710,8 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
         self.callback_handler.optimizer = self.optimizer
         self.callback_handler.lr_scheduler = self.lr_scheduler
         self.callback_handler.train_dataloader = train_dataloader
+        trial_name = self.hp_name(trial) if self.hp_name is not None else None
+        self.callback_handler.trial_info = {"trial": self._trial, "trial_name": trial_name}
```
Comment (sgugger) [inline src/transformers/trainer.py:714]: Should this be in the state instead?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -823,7 +833,21 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
 
         return TrainOutput(self.state.global_step, tr_loss.item() / self.state.global_step)
 
-    def _maybe_log_save_evalute(self, tr_loss, model, trial, epoch):
+    def _rewrite_metrics(self, prefix, metrics):
```
Comment (sgugger) [inline src/transformers/trainer.py:836]: This does not sound very good. Lots of behaviors and tests rely on the metric names, so I'd avoid any renaming in the main Trainer.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -836,12 +860,13 @@ def _maybe_log_save_evalute(self, tr_loss, model, trial, epoch):
             )
             self._logging_loss_scalar = tr_loss_scalar
 
-            self.log(logs)
+            self.log("train", logs)
```
Comment (sgugger) [inline src/transformers/trainer.py:863]: This will break the `NotebookProgressCallback` at least, and perhaps other things. Really not in favor of renaming things inside the main Trainer as I said earlier.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -902,13 +929,18 @@ def _save_checkpoint(self, model, trial, metrics=None):
         if self.is_world_process_zero():
             self._rotate_checkpoints(use_mtime=True)
 
+    def _run_name(self, trial):
```
Comment (sgugger) [inline src/transformers/trainer.py:932]: This seems to be only used once, not sure we need the refactor in a separate function for just two lines?

File: src/transformers/trainer_utils.py [hunk 6]
```diff
@@ -110,10 +111,15 @@ def default_compute_objective(metrics: Dict[str, float]) -> float:
     Return:
         :obj:`float`: The objective to minimize or maximize
     """
+    metrics = copy.deepcopy(metrics)
     loss = metrics.pop("eval_loss", None)
     _ = metrics.pop("epoch", None)
     _ = metrics.pop("total_flos", None)
-    return loss if len(metrics) == 0 else sum(metrics.values())
+    if len(metrics) != 0:
+        raise RuntimeError(
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:119]: Nice error!

File: tests/test_trainer.py [hunk 7]
```diff
@@ -143,6 +153,7 @@ def get_regression_trainer(a=0, b=0, double_output=False, train_len=64, eval_len
         optimizers = kwargs.pop("optimizers", (None, None))
         output_dir = kwargs.pop("output_dir", "./regression")
         args = TrainingArguments(output_dir, **kwargs)
+        trainer_args = trainer_args or {}
```
Comment (sgugger) [inline tests/test_trainer.py:156]: Can we pop from the kwargs instead (like for optimizers just above?)

File: tests/test_trainer.py [hunk 8]
```diff
@@ -617,3 +629,48 @@ def assert_flos_extraction(trainer, wrapped_model_to_check):
...
+                a = 0
+                b = 0
+            config = RegressionModelConfig(a=a, b=b, double_output=False)
+
+            return RegressionPreTrainedModel(config)
+
+        def hp_name(trial):
+            return MyTrialShortNamer.shortname(trial.params)
+
+        trainer_args = dict(model_init=model_init)
```
Comment (sgugger) [inline tests/test_trainer.py:663]: If popped from the kwargs above, we can just pass model_init below, which is easier.

**Review by sgugger — COMMENTED** [review]
I don't think changing the metrics names in the logs is a good idea as there are other components relying on them. Renaming the keys inside the TensorBoardCallback to make tensorboard happier seems fine but outside of it, I would prefer to avoid any breaking change.

File: src/transformers/testing_utils.py (hunk 1 above)
Comment (madlag) [reply src/transformers/testing_utils.py:237]: OK.

File: src/transformers/trainer.py (hunk 2 above)
Comment (madlag) [reply src/transformers/trainer.py:714]: Yes, did not want to change the state, to limit the amount of changes, but I can move it to the state.

File: src/transformers/trainer.py (hunk 3 above)
Comment (madlag) [reply src/transformers/trainer.py:836]: I did not change the eval part, for this reason. But I was not aware of same issues with train metrics. I can move code, but I will have to change callbacks to send more information to do the same job in the integrations (not really a problem of course).
Another way would be to change the tests, as prefixing metrics with eval_ is not really easy to handle, as metrics name themselves may contain underscores, whereas splitting by / would make it easier to handle different metrics.
But as you said, if a lot of behaviors are relying on this names, and not just tests, it's hard to change everywhere.
The eternal problem between evolving code and keeping backward-compatibility...

File: src/transformers/trainer.py (hunk 4 above)
Comment (madlag) [reply src/transformers/trainer.py:863]: The issue there is that there is no way to know if we are training or evaluating in the log function, so I won't be able to do anything for the TensorBoard or Wandb integration.

File: tests/test_trainer.py (hunk 7 above)
Comment (madlag) [reply tests/test_trainer.py:156]: Yes, of course.

File: tests/test_trainer.py (hunk 8 above)
Comment (madlag) [reply tests/test_trainer.py:663]: OK,

File: src/transformers/trainer.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/trainer.py:863]: I don't understand that part. Except for the final evaluation, we're always in the middle of training.

File: tests/test_trainer.py (hunk 8 above)
Comment (madlag) [reply tests/test_trainer.py:663]: OK, but I am not really fan of poping stuff from dicts.
Here we have to know exactly which parameters are going to the trainer, and which are going to the trainer arguments.
If you forget one argument, or if two have the same name, it's becoming a mess.
For example, the RuntimeError I raised in default_compute_objective was because someone added "total_flos" to the metrics somewhere else but did not realize that default_compute_objective was popping values out of the metrics dictionary, and adding everything that was left in it. 
So the final eval metric in the end was the sum of the true eval metric and "total_flos", not really good.
So, yes I can use a dict.pop , but it makes some brittle code IMHO.

Comment (sgugger) [reply tests/test_trainer.py:663]: I don't really see why this is similar to "total_flos" being added to the metrics (and the conclusion there is that it should never have been added to the metrics cause it's not one and it's polluting the logs when only the final value matters).

It's a helper function for some tests so there is little point debating on its API for hours. I was just expressing the fact I would have liked some consistency in this test file on how the defaults are changed but if it irks you that much, do what you want.

File: src/transformers/testing_utils.py (hunk 1 above)
Comment (madlag) [reply src/transformers/testing_utils.py:237]: Is there something I can do to add optuna / ray test cases on the CI with the proper configuration ?

File: src/transformers/trainer.py (hunk 2 above)
Comment (madlag) [reply src/transformers/trainer.py]: Resolved in https://github.com/huggingface/transformers/pull/7935/commits/3179a1f8044446fd89275a9fbde82a2fa4447ce1

File: tests/test_trainer.py (hunk 8 above)
Comment (madlag) [reply tests/test_trainer.py]: Fixed in https://github.com/huggingface/transformers/pull/7935/commits/1c4d60817d922783f8bbac1764674918476dcbd9 .

File: tests/test_trainer.py (hunk 7 above)
Comment (madlag) [reply tests/test_trainer.py]: Fixed in https://github.com/huggingface/transformers/pull/7935/commits/1c4d60817d922783f8bbac1764674918476dcbd9

File: src/transformers/trainer.py (hunk 5 above)
Comment (madlag) [reply src/transformers/trainer.py]: Solved in https://github.com/huggingface/transformers/pull/7935/commits/d83d80d075b5dd7278020c93a3fb0235a3405ecc

File: src/transformers/trainer.py (hunk 4 above)
Comment (madlag) [reply src/transformers/trainer.py]: Not for the majority of metrics, but for example total_flos and epoch are added "manually" in the log() function, and they are not prefixed by "train", "eval" or anything, so there is a name collision there. And if somebody add other stuff in the log function, the collision will affect more data.

Comment (madlag) [reply src/transformers/trainer.py]: Fixed in https://github.com/huggingface/transformers/pull/7935/commits/d57d45fe6246e4d6f85a7dc683858546da569e18

File: src/transformers/trainer.py (hunk 3 above)
Comment (madlag) [reply src/transformers/trainer.py]: Fixed in https://github.com/huggingface/transformers/pull/7935/commits/d57d45fe6246e4d6f85a7dc683858546da569e18

File: src/transformers/testing_utils.py (hunk 1 above)
Comment (madlag) [reply src/transformers/testing_utils.py:237]: OK, so I will have a look how to integrate this.

**Review by sgugger — APPROVED** [review]
Thanks for fixing/making all of this better!
